ext/civi_member/Civi/Api4/Action/Membership/Validate.php

```diff
@@ -0,0 +1,86 @@
+<?php
+
+/*
+ +--------------------------------------------------------------------+
+ | Copyright CiviCRM LLC. All rights reserved.                        |
+ |                                                                    |
+ | This work is published under the GNU AGPLv3 license with some      |
+ | permitted exceptions and without any warranty. For full license    |
+ | and copyright information, see https://civicrm.org/licensing       |
+ +--------------------------------------------------------------------+
+ */
+
+namespace Civi\Api4\Action\Membership;
+
+use Civi\Api4\Generic\ValidateAction;
+use Civi\Api4\Event\ValidateValuesEvent;
+
+/**
+ * Validate membership parameters before creating/updating Memberships.
+ */
+class Validate extends ValidateAction {
+
+  protected function onValidateValues(ValidateValuesEvent $e) {
```

Comment:
There is quite a lot of validation in here and a good chance that someone will want to customise/disable some of it at some point. Can it be implemented as an event subscriber instead using IsActiveTrait so that it can be disabled/replaced if required?

Reply:
@mattwire good point. I learnt from our discussion in dev-financial and core team that its better to have this rules in more central and accessible place maybe at BAO level instead of declaring all those rules in APIv4 validate action and the former could act as a wrapper to host listeners. I like the idea of https://github.com/civicrm/civicrm-core/pull/30716#discussion_r1696994410 . At this point I am thinking of migrating the rules in a separate BAO fns instead of APIv4 validate action supplemented with unit tests, esp. when the generic entity's validate action's signature is yet to be finalized and that should also work for pseudo entity like Order, payment etc. (and hookable)

Reply:
@monishdeb & @mattwire - I think it would be good to write the tests around the api layer & then look to migrating to a listener layer - that might sit in the BAO. As long as there are tests and we have not created any contact we can't maintain (e.g listeners with a non-finalised signature) we can move stuff around once the tests are in place. 

I feel like the api signature is probably agreed, or close to so it would be good to focus this PR on that (since I think the goal is also to switch the form layer to use it, at least to a limited extent) and that work would be agnostic of everything except the api signature.

The one part we haven't tied down is how the ability to set which validations apply is set - although that might be part of that next chunk.

One reason I'm trying to break it up is that I'm aware that several people are doing parts of this work & trying to see where work does not have to be blocked by other work - ie any attempt to move this to the BAO before that signature is finalised (which will likely take longer) probably is not ultimately helpful - but getting the test/s merged will be very helpful for getting that validate structure code done